pkg/maps/bwmap/cell.go

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package bwmap
+
+import (
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/statedb"
+	"github.com/cilium/statedb/reconciler"
+
+	"github.com/cilium/cilium/pkg/bpf"
+	"github.com/cilium/cilium/pkg/datapath/types"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+// Cell manages the cilium_throttle BPF map for implementing per-endpoint
+// bandwidth management. The cell provides RWTable[Edt] to which per
+// endpoint bandwidth limits can be inserted. Use [NewEdt] to create the
+// object. The table can be inspected with "cilium-dbg statedb bandwidth-edts".
+// A reconciler is registered that reconciles the table with the cilium_throttle
+// map.
+var Cell = cell.Module(
+	"bwmap",
+	"Manages the endpoint bandwidth limit BPF map",
+
+	cell.Provide(
+		NewEdtTable,
+		statedb.RWTable[Edt].ToTable,
+		newThrottleMap,
+	),
+	cell.ProvidePrivate(
+		edtReconcilerConfig,
+	),
+	cell.Invoke(
+		statedb.RegisterTable[Edt],
+		registerReconciler,
+		bpf.RegisterTablePressureMetricsJob[Edt, throttleMap],
+	),
+)
+
+func registerReconciler(cfg types.BandwidthConfig, rcfg reconciler.Config[Edt], params reconciler.Params) error {
+	if cfg.EnableBandwidthManager {
+		return reconciler.Register(rcfg, params)
+	}
+	return nil
+}
+
+func edtReconcilerConfig(m throttleMap, edts statedb.RWTable[Edt]) reconciler.Config[Edt] {
+	ops := bpf.NewMapOps[Edt](m.Map)
+
+	return reconciler.Config[Edt]{
+		FullReconcilationInterval: time.Hour,
+		RetryBackoffMinDuration:   100 * time.Millisecond,
+		RetryBackoffMaxDuration:   time.Minute,
+		IncrementalRoundSize:      100,
```

Comment:
How are these values determined? Is there a way to monitor the effectiveness of these values so they can be tuned if needed?

Reply:
Good question! There's metrics for both the operations and for statedb itself which would give a way of understanding the behavior in a busy cluster. Currently the values are just chosen to be somewhat reasonable for the bandwidth map, but of course it's just a guess.

I've thought a bit that perhaps we should make all these reconciler settings runtime configurable (without need to opt-in for it). It wouldn't be too hard to implement and then we'd have a way to tune these if the need arises...

Reply:
Another point of view is that we expect the BPF map operations to be very reliable and the usual failure scenario is that the BPF map is full. So having a fairly long max backoff makes sense for that to avoid busy work.

The incremental round size is there for amortizing the cost of write transactions and to avoid waking up code watching the objects too often. In benchmarks (https://github.com/cilium/statedb/tree/main/reconciler/benchmark) I've found that 100 is pretty much the sweet spot usually. 

The full reconciliation interval is a bit arbitrary. It marks how often we resynchronize (e.g. call Update() on each object) and how often to Prune() to remove unexpected/stale entries (usually relevant when just starting). Having it very long would mean we don't recover from outside interference quickly, but having it too short means we'll spend time doing unnecessary work. Probably 1 hour is too long for the general case.

Reply:
Runtime reconfigurability would definitely be a great next step. Thanks for explaining the purpose of each value and the reasoning behind the values! Helps a lot to understand how it all works.